domain/port/state/state.go

```diff
@@ -0,0 +1,282 @@
+	if err != nil {
+		return errors.Annotate(err, "preparing get unit endpoints statement")
+	}
+
+	insertUnitEndpoint, err := st.Prepare("INSERT INTO unit_endpoint (*) VALUES ($unitEndpoint.*)", unitEndpoint{})
+	if err != nil {
+		return errors.Annotate(err, "preparing insert unit endpoint statement")
+	}
+
+	getProtocols, err := st.Prepare("SELECT &protocol.* FROM protocol", protocol{})
+	if err != nil {
+		return errors.Annotate(err, "preparing get protocol ID statement")
+	}
+
+	getOpenedPortsOnEndpoints, err := st.Prepare(`
+SELECT &endpointPortRange.*
+FROM port_range
+JOIN protocol ON port_range.protocol_id = protocol.id
+JOIN unit_endpoint ON port_range.unit_endpoint_uuid = unit_endpoint.uuid
+WHERE unit_endpoint.unit_uuid = $unitUUID.unit_uuid
+AND unit_endpoint.endpoint IN ($endpoints[:])
+`, endpointPortRange{}, unitUUID, endpointsUnderAction)
+	if err != nil {
+		return errors.Annotate(err, "preparing get opened ports on endpoints statement")
+	}
+
+	insertPortRange, err := st.Prepare("INSERT INTO port_range (*) VALUES ($unitPortRange.*)", unitPortRange{})
+	if err != nil {
+		return errors.Annotate(err, "preparing insert port range statement")
+	}
+
+	clearPortRange, err := st.Prepare("DELETE FROM port_range WHERE unit_endpoint_uuid IN ($endpointUUIDs[:]) ", endpointUUIDs{})
+	if err != nil {
+		return errors.Annotate(err, "preparing clear port range statement")
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+
+		// Get the endpoints under action which currently exist in state.
+		// Cache the endpoint's UUIDs for later use.
+		var currenUnitEndpoint []endpoint
+		err = tx.Query(ctx, getUnitEndpoints, unitUUID, endpointsUnderAction).GetAll(&currenUnitEndpoint)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			currenUnitEndpoint = []endpoint{}
+		} else if err != nil {
+			return domain.CoerceError(err)
+		}
+
+		foundEndpoints := set.NewStrings()
+		endpointsUnderActionToUUID := map[string]endpointUUID{}
+		for _, ep := range currenUnitEndpoint {
+			foundEndpoints.Add(ep.Endpoint)
+			endpointsUnderActionToUUID[ep.Endpoint] = endpointUUID{UUID: ep.UUID}
+		}
+
+		// Insert any new endpoints that are required.
+		requiredEndpoints := set.NewStrings(endpointsUnderAction...).Difference(foundEndpoints)
+		newUnitEndpoints := make([]unitEndpoint, requiredEndpoints.Size())
+		for i, endpoint := range requiredEndpoints.Values() {
+			uuid, err := uuid.NewUUID()
+			if err != nil {
+				return errors.Annotatef(err, "generating UUID for unit endpoint")
+			}
+			newUnitEndpoints[i] = unitEndpoint{
+				UUID:     uuid.String(),
+				UnitUUID: unit,
+				Endpoint: endpoint,
+			}
+			endpointsUnderActionToUUID[endpoint] = endpointUUID{UUID: uuid.String()}
+		}
+
+		if len(newUnitEndpoints) > 0 {
+			err = tx.Query(ctx, insertUnitEndpoint, newUnitEndpoints).Run()
+			if err != nil {
+				return domain.CoerceError(err)
+			}
+		}
+
+		// Construct a map of protocol names to protocol IDs for later use.
```

Comment:
Could you not just get this out in the `getOpenedPortsOnEndpoints` query? You are already getting out the protocol name and joining with the table. It only seems to be used to look up the name to get the ID.

In fact, do we even need the name? Can we get away with just the ID?

Reply:
Not necessarily. We can't say that our query will actually return anything out of the database if no ports are opened. When we later then try to insert a row, we won't know why id to insert.

There probably is a very clever bit of SQLair I could write that looks up the protocol id in flight + inserts into the table. Let's chat tomorrow about it maybe

Perhaps changing straight over to the protcol name as primary key. This could be good then actually. protocol name have to be unique anyway!

Reply:
Simply getting away with id wouldn't really be a good idea. it results in a "repeating group"

https://www.freecodecamp.org/news/database-normalization-1nf-2nf-3nf-table-examples/

Reply:
Had a think about this, and I really can't think of any way to get the IDs out with it also being a bulk insert. It would be doable with it being a regular insert (you could do `INSERT INTO ... SELECT ...`) but that won't work right with a bulk

Reply:
Just having a separate query is fine by me, though somehow it feels wasteful since these will never change. I wonder if theres a good reason not to load these into the state struct on initialization or something? (premature optimization might be one)